ElasticSearch sink losing some log lines

We are currently seeing some logs missing and notice errors like:

```cernan-g4c4t cernan [elastic::client::requests][174][2017-11-30T18:18:45.299068595+00:00][ERROR] Elasticsearch Response: correlation_id: 'be20d167-e875-4f0d-9399-b2546ddb3e81', error: 'https://search-logs55-g4fexgk6iiyjifhfmiuykc6pxu.us-west-2.es.amazonaws.com/_bulk: failed to lookup address information: Try again'
cernan-g4c4t cernan [cernan::sink::elasticsearch][243][2017-11-30T18:18:45.299097051+00:00][ERROR] Unable to write, Reqwest Client Error: failed to lookup address information: Try again
```

@blt verified that it's possible for ES to signal that the majority of items were written but not all.

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -7,7 +7,7 @@ license = "MIT"
 name = "cernan"
 readme = "README.md"
 repository = "https://github.com/postmates/cernan"
-version = "0.8.2"
+version = "0.8.3-pre"
 
 [[bin]]
 name = "cernan"

--- a/src/config.rs
+++ b/src/config.rs
@@ -568,6 +568,14 @@ pub fn parse_config_file(buffer: &str, verbosity: u64) -> Args {
             let mut res = ElasticsearchConfig::default();
             res.config_path = Some("sinks.elasticsearch".to_string());
 
+            res.delivery_attempt_limit = snk.get("delivery_attempt_limit")
+                .map(|p| {
+                    p.as_integer()
+                        .expect("could not parse sinks.elasticsearch.delivery_attempt_limit")
+                        as u8
+                })
+                .unwrap_or(res.delivery_attempt_limit);
+
             res.port = snk.get("port")
                 .map(|p| {
                     p.as_integer()
@@ -1083,6 +1091,7 @@ scripts-directory = "/foo/bar"
   index-prefix = "prefix-"
   secure = true
   flush_interval = 2020
+  delivery_attempt_limit = 33
 "#;
 
         let args = parse_config_file(config, 4);
@@ -1095,6 +1104,7 @@ scripts-directory = "/foo/bar"
         assert_eq!(es.index_prefix, Some("prefix-".into()));
         assert_eq!(es.secure, true);
         assert_eq!(es.flush_interval, 2020);
+        assert_eq!(es.delivery_attempt_limit, 33);
     }
 
     #[test]

--- a/src/sink/elasticsearch.rs
+++ b/src/sink/elasticsearch.rs
@@ -200,15 +200,10 @@ impl Sink for Elasticsearch {
                         Ok(item) => {
                             let uuid = uuid::Uuid::parse_str(item.id())
                                 .expect("catastrophic error, TID not a UUID");
-                            match self.buffer
+                            if let Ok(idx) = self.buffer
                                 .binary_search_by(|probe| probe.uuid.cmp(&uuid))
                             {
-                                Ok(idx) => {
-                                    self.buffer.remove(idx);
-                                }
-                                Err(_) => {
-                                    unreachable!();
-                                }
+                                self.buffer.remove(idx);
                             }
                             ELASTIC_RECORDS_TOTAL_DELIVERED
                                 .fetch_add(1, Ordering::Relaxed);
@@ -216,17 +211,12 @@ impl Sink for Elasticsearch {
                         Err(item) => {
                             let uuid = uuid::Uuid::parse_str(item.id())
                                 .expect("catastrophic error, TID not a UUID");
-                            match self.buffer
+                            if let Ok(idx) = self.buffer
                                 .binary_search_by(|probe| probe.uuid.cmp(&uuid))
                             {
-                                Ok(idx) => {
-                                    self.buffer[idx].attempts += 1;
-                                    if self.buffer[idx].attempts > self.delivery_attempt_limit {
-                                        self.buffer.remove(idx);
-                                    }
-                                }
-                                Err(_) => {
-                                    unreachable!();
+                                self.buffer[idx].attempts += 1;
+                                if self.buffer[idx].attempts > self.delivery_attempt_limit {
+                                    self.buffer.remove(idx);
                                 }
                             }
                             ELASTIC_RECORDS_TOTAL_FAILED

--- a/src/sink/elasticsearch.rs
+++ b/src/sink/elasticsearch.rs
@@ -13,7 +13,7 @@ use std::error::Error;
 use std::sync;
 use std::sync::Arc;
 use std::sync::atomic::{AtomicUsize, Ordering};
-use uuid::Uuid;
+use uuid;
 
 lazy_static! {
     /// Total deliveries made
@@ -94,7 +94,7 @@ impl Default for ElasticsearchConfig {
 ///
 /// Refer to the documentation on `ElasticsearchConfig` for more details.
 pub struct Elasticsearch {
-    buffer: Vec<LogLine>,
+    buffer: Vec<(uuid::Uuid, LogLine)>,
     secure: bool,
     host: String,
     port: usize,
@@ -123,10 +123,11 @@ impl Elasticsearch {
         assert!(!self.buffer.is_empty());
         use serde_json::{to_string, Value};
         for m in &self.buffer {
-            let uuid = Uuid::new_v4().hyphenated().to_string();
+            let uuid = m.0.hyphenated().to_string();
+            let line = &m.1;
             let header: Value = json!({
                 "index": {
-                    "_index" : idx(&self.index_prefix, m.time),
+                    "_index" : idx(&self.index_prefix, line.time),
                     "_type" : self.index_type.clone(),
                     "_id" : uuid.clone(),
                 }
@@ -135,15 +136,15 @@ impl Elasticsearch {
             buffer.push('\n');
             let mut payload: Value = json!({
                 "uuid": uuid,
-                "path": m.path.clone(),
-                "payload": m.value.clone(),
-                "timestamp": format_time(m.time),
+                "path": line.path.clone(),
+                "payload": line.value.clone(),
+                "timestamp": format_time(line.time),
             });
             let obj = payload.as_object_mut().unwrap();
-            for &(ref k, ref v) in m.tags.iter() {
+            for &(ref k, ref v) in line.tags.iter() {
                 obj.insert(k.clone(), Value::String(v.clone()));
             }
-            for &(ref k, ref v) in m.fields.iter() {
+            for &(ref k, ref v) in line.fields.iter() {
                 obj.insert(k.clone(), Value::String(v.clone()));
             }
             buffer.push_str(&to_string(&obj).unwrap());
@@ -169,7 +170,6 @@ impl Sink for Elasticsearch {
 
         let mut buffer = String::with_capacity(4048);
         self.bulk_body(&mut buffer);
-        debug!("BODY: {:?}", buffer);
         let bulk_resp: Result<BulkResponse> = client
             .request(BulkRequest::new(buffer))
             .send()
@@ -178,11 +178,22 @@ impl Sink for Elasticsearch {
 
         match bulk_resp {
             Ok(bulk) => {
-                self.buffer.clear();
                 ELASTIC_RECORDS_DELIVERY.fetch_add(1, Ordering::Relaxed);
                 for item in bulk.iter() {
                     match item {
-                        Ok(_item) => {
+                        Ok(item) => {
+                            let uuid = uuid::Uuid::parse_str(item.id())
+                                .expect("catastrophic error, TID not a UUID");
+                            match self.buffer
+                                .binary_search_by(|probe| probe.0.cmp(&uuid))
+                            {
+                                Ok(idx) => {
+                                    self.buffer.remove(idx);
+                                }
+                                Err(_) => {
+                                    unreachable!();
+                                }
+                            }
                             ELASTIC_RECORDS_TOTAL_DELIVERED
                                 .fetch_add(1, Ordering::Relaxed);
                         }
@@ -286,7 +297,8 @@ impl Sink for Elasticsearch {
 
     fn deliver_line(&mut self, mut lines: sync::Arc<Option<LogLine>>) -> () {
         let line: LogLine = sync::Arc::make_mut(&mut lines).take().unwrap();
-        self.buffer.push(line);
+        let uuid = uuid::Uuid::new_v4();
+        self.buffer.push((uuid, line));
     }
 
     fn valve_state(&self) -> Valve {

--- a/src/sink/elasticsearch.rs
+++ b/src/sink/elasticsearch.rs
@@ -18,6 +18,8 @@ use uuid;
 lazy_static! {
     /// Total deliveries made
     pub static ref ELASTIC_RECORDS_DELIVERY: Arc<AtomicUsize> = Arc::new(AtomicUsize::new(0));
+    /// Total internal buffer entries
+    pub static ref ELASTIC_INTERNAL_BUFFER_LEN: Arc<AtomicUsize> = Arc::new(AtomicUsize::new(0));
     /// Total records delivered in the last delivery
     pub static ref ELASTIC_RECORDS_TOTAL_DELIVERED: Arc<AtomicUsize> = Arc::new(AtomicUsize::new(0));
     /// Total records that failed to be delivered due to error
@@ -67,7 +69,11 @@ pub struct ElasticsearchConfig {
     pub index_type: String,
     /// Determines whether to use HTTP or HTTPS when publishing to
     /// Elasticsearch.
-    pub secure: bool, // whether http or https
+    pub secure: bool, 
+    /// Determine how many times to attempt the delivery of a log line before
+    /// dropping it from the buffer. Failures of a global bulk request does not
+    /// count against this limit.
+    pub delivery_attempt_limit: u8,
     /// The Elasticsearch host. May be an IP address or DNS hostname.
     pub host: String,
     /// The Elasticsearch port.
@@ -84,20 +90,28 @@ impl Default for ElasticsearchConfig {
             host: "127.0.0.1".to_string(),
             index_prefix: None,
             index_type: "payload".to_string(),
+            delivery_attempt_limit: 10,
             port: 9200,
             flush_interval: 1,
         }
     }
 }
 
+struct Line {
+    attempts: u8,
+    uuid: uuid::Uuid,
+    line: LogLine,
+}
+
 /// The elasticsearch sink struct.
 ///
 /// Refer to the documentation on `ElasticsearchConfig` for more details.
 pub struct Elasticsearch {
-    buffer: Vec<(uuid::Uuid, LogLine)>,
+    buffer: Vec<Line>,
     secure: bool,
     host: String,
     port: usize,
+    delivery_attempt_limit: u8,
     index_prefix: Option<String>,
     index_type: String,
     flush_interval: u64,
@@ -115,6 +129,7 @@ impl Elasticsearch {
             port: config.port,
             index_prefix: config.index_prefix,
             index_type: config.index_type,
+            delivery_attempt_limit: config.delivery_attempt_limit,
             flush_interval: config.flush_interval,
         }
     }
@@ -123,8 +138,8 @@ impl Elasticsearch {
         assert!(!self.buffer.is_empty());
         use serde_json::{to_string, Value};
         for m in &self.buffer {
-            let uuid = m.0.hyphenated().to_string();
-            let line = &m.1;
+            let uuid = m.uuid.hyphenated().to_string();
+            let line = &m.line;
             let header: Value = json!({
                 "index": {
                     "_index" : idx(&self.index_prefix, line.time),
@@ -176,6 +191,7 @@ impl Sink for Elasticsearch {
             .unwrap()
             .into_response::<BulkResponse>();
 
+        ELASTIC_INTERNAL_BUFFER_LEN.store(self.buffer.len(), Ordering::Relaxed);
         match bulk_resp {
             Ok(bulk) => {
                 ELASTIC_RECORDS_DELIVERY.fetch_add(1, Ordering::Relaxed);
@@ -185,7 +201,7 @@ impl Sink for Elasticsearch {
                             let uuid = uuid::Uuid::parse_str(item.id())
                                 .expect("catastrophic error, TID not a UUID");
                             match self.buffer
-                                .binary_search_by(|probe| probe.0.cmp(&uuid))
+                                .binary_search_by(|probe| probe.uuid.cmp(&uuid))
                             {
                                 Ok(idx) => {
                                     self.buffer.remove(idx);
@@ -198,6 +214,21 @@ impl Sink for Elasticsearch {
                                 .fetch_add(1, Ordering::Relaxed);
                         }
                         Err(item) => {
+                            let uuid = uuid::Uuid::parse_str(item.id())
+                                .expect("catastrophic error, TID not a UUID");
+                            match self.buffer
+                                .binary_search_by(|probe| probe.uuid.cmp(&uuid))
+                            {
+                                Ok(idx) => {
+                                    self.buffer[idx].attempts += 1;
+                                    if self.buffer[idx].attempts > self.delivery_attempt_limit {
+                                        self.buffer.remove(idx);
+                                    }
+                                }
+                                Err(_) => {
+                                    unreachable!();
+                                }
+                            }
                             ELASTIC_RECORDS_TOTAL_FAILED
                                 .fetch_add(1, Ordering::Relaxed);
                             if let Some(cause) = item.cause() {
@@ -298,7 +329,7 @@ impl Sink for Elasticsearch {
     fn deliver_line(&mut self, mut lines: sync::Arc<Option<LogLine>>) -> () {
         let line: LogLine = sync::Arc::make_mut(&mut lines).take().unwrap();
         let uuid = uuid::Uuid::new_v4();
-        self.buffer.push((uuid, line));
+        self.buffer.push(Line { uuid: uuid, line: line, attempts: 0 });
     }
 
     fn valve_state(&self) -> Valve {

--- a/src/source/internal.rs
+++ b/src/source/internal.rs
@@ -157,6 +157,12 @@ impl Source for Internal {
                     self.chans
                 );
                 atom_non_zero_telem!(
+                    "cernan.sinks.elasticsearch.internal.buffer_len",
+                    sink::elasticsearch::ELASTIC_INTERNAL_BUFFER_LEN,
+                    self.tags,
+                    self.chans
+                );
+                atom_non_zero_telem!(
                     "cernan.sinks.elasticsearch.records.total_delivered",
                     sink::elasticsearch::ELASTIC_RECORDS_TOTAL_DELIVERED,
                     self.tags,

